### PR TITLE
feature/add-example-scripts

### PR DESCRIPTION
**Pull request by gregjohnso**

### CODE REVIEW

**Comment by counkomol** [comment]
Please remake this PR as a merge into pytorch_fnet:master.

File: .environment.yml [hunk 1]
```diff
@@ -11,10 +11,12 @@ dependencies:
...
-  - pytorch=0.4
+  - pytorch=1.0.1
   - scipy
   - tifffile
   - torchvision
   - tqdm
+  - scikit-image
   - pip:
     - czifile==0.1.5
+    - t4
```
Comment (JacksonMaxfield) [inline .environment.yml]: You can split up package dependencies in `setup.py` now!
Example from [aics-deformation `setup.py`](https://github.com/AllenCellModeling/aics-deformation/blob/master/setup.py#L31)

File: examples/download_and_train.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+len(aics_pipeline['fov'])
+
+image_ids = [k for k in aics_pipeline['fov']][0:n_images_to_download]
+
+metadata = {}
+for image_id in image_ids:
+    metadata[image_id] = aics_pipeline['fov'][image_id].get_meta()
+
+image_save_paths = ['{}/{}'.format(image_save_dir, image_id) for image_id in image_ids]
```
Comment (JacksonMaxfield) [inline examples/download_and_train.py]: With `pathlib.Path`s this becomes:
`image_save_paths = [image_save_dir / image_id for image_id in image_ids`

But again, metadata storage and remaking the files list wont be necessary once the `t4` fetch update PR goes through.

File: examples/download_and_train.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    except:
+        pass
+    
+###################################################
+# Make a manifest of all of the files in csv form
+###################################################
+
+
+
+df = pd.DataFrame(columns = ['path_tiff','channel_signal', 'channel_target'])
```
Comment (JacksonMaxfield) [inline examples/download_and_train.py]: Does `flake8` not care about spaces before and after params?
My IDE gives me linter warnings when param values are with spaces.
`df = pd.DataFrame(columns=['path_tiff','channel_signal', 'channel_target'])`

File: examples/download_and_train.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+df['path_tiff'] = image_save_paths
+
+df['channel_signal'] = [metadata[image_id]['content_info']['brighfield_channel'] for image_id in image_ids]
+df['channel_target'] = [metadata[image_id]['content_info']['dna_channel'] for image_id in image_ids]
```
Comment (JacksonMaxfield) [inline examples/download_and_train.py]: Nit: I would prefer this dataframe initialization all as one.

```
rows = [{
   "path_tiff": image_save_paths / image_id,
   "channel_signal": metadata[image_id]["content_info"]["brighfield_channel"],
   "channel_target": metadata[image_id]["content_info"]["dna_channel"]
} for image_id in image_ids]

df = pd.DataFrame(rows)
```

File: examples/download_and_train.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+# Make a manifest of all of the files in csv form
+###################################################
+
+
+
+df = pd.DataFrame(columns = ['path_tiff','channel_signal', 'channel_target'])
+
+df['path_tiff'] = image_save_paths
+
+df['channel_signal'] = [metadata[image_id]['content_info']['brighfield_channel'] for image_id in image_ids]
```
Comment (JacksonMaxfield) [inline examples/download_and_train.py]: Note to self: I really need to fix `brighfield_channel` with `brightfield_channel`

File: examples/download_and_train.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+df['channel_target'] = [metadata[image_id]['content_info']['dna_channel'] for image_id in image_ids]
+
+n_train_images = int(n_images_to_download*train_fraction)
+df_train = df[0:n_train_images]
```
Comment (JacksonMaxfield) [inline examples/download_and_train.py]: No need for `0` on slice.

File: examples/download_and_train.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+prefs['dataset_val'] = 'fnet.data.AICSTiffDataset'
+prefs['dataset_val_kwargs'] = {'path_csv': data_save_path_test}
+
+
+with open(prefs_save_path, 'w') as fp:
+    json.dump(prefs, fp)
+
+command_str = 'python ../fnet/cli/train_model.py {} --gpu_ids {}'.format(prefs_save_path, gpu_id)
+
+print(command_str)
```
Comment (JacksonMaxfield) [inline examples/download_and_train.py]: Is `print` really needed?

File: fnet/cli/train_model.py [hunk 8]
```diff
@@ -114,6 +114,9 @@ def main(args: Optional[argparse.Namespace] = None) -> None:
     print('*** Training options ***')
     pprint.pprint(vars(args))
 
+    if not os.path.exists(args.path_save_dir):
```
Comment (JacksonMaxfield) [inline fnet/cli/train_model.py]: `Path.mkdir(exist_ok=True)`

File: fnet/data/__init__.py [hunk 9]
```diff
@@ -1,6 +1,7 @@
 from fnet.data.bufferedpatchdataset import BufferedPatchDataset
 from fnet.data.tiffdataset import TiffDataset
 from fnet.data.fnetdataset import FnetDataset
+from fnet.data.aicstiffdataset import AICSTiffDataset
 
 
-__all__ = ['BufferedPatchDataset', 'FnetDataset', 'TiffDataset']
+__all__ = ['BufferedPatchDataset', 'FnetDataset', 'TiffDataset', 'AICSTiffDataset']
```
Comment (JacksonMaxfield) [inline fnet/data/__init__.py]: I didn't know this is how you get rid of those `flake8` linter warnings. Neat.

File: fnet/data/aicstiffdataset.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
+import torch.utils.data
```
Comment (JacksonMaxfield) [inline fnet/data/aicstiffdataset.py]: @counkomol What would be preferred method of comparing this to `TiffDataset` because you thought these files were redundant.

File: setup.py [hunk 11]
```diff
@@ -21,8 +21,10 @@
         'pandas',
         'scipy',
         'tifffile',
-        'torch==0.4',
+        'torch==1.0.1',
         'tqdm',
+        'scikit-image',
+        't4'
```
Comment (JacksonMaxfield) [inline setup.py]: As discussed above, move these to an `extras_require` block.

File: fnet/data/aicstiffdataset.py (hunk 10 above)
Comment (counkomol) [reply fnet/data/aicstiffdataset.py]: I believe the examples can be redone using TiffDataset.

Comment (counkomol) [inline fnet/data/aicstiffdataset.py]: What functionality does this provide over TiffDataset?

File: setup.py [hunk 12]
```diff
@@ -21,8 +21,10 @@
...
     ],
     extras_require={
```
Comment (counkomol) [inline setup.py]: Please add the example specific dependencies here.

File: examples/download_and_train.py (hunk 3 above)
Comment (counkomol) [inline examples/download_and_train.py]: Because new Python code is added, the `flake8` check should be updated to analyze the `./examples` folder.

File: fnet/cli/predict.py [hunk 13]
```diff
@@ -93,6 +93,7 @@ def item_from_dataset(
     else:
         item = dataset[idx]
     target = None
+
```
Comment (counkomol) [inline fnet/cli/predict.py]: Why?

**Comment by counkomol** [comment]
Also, please resolve any merge conflicts.

File: examples/download_and_train.py (hunk 3 above)
Comment (counkomol) [reply examples/download_and_train.py]: To be more explicit, you can update the Makefile like so:
```
check:
     flake8 --extend-ignore=E501 --max-complexity 10 --exclude=tests/ignore/* fnet tests examples
```

Also, in the other outstanding PR, the Travis CI test was updated to include flake8 checking.

File: fnet/cli/train_model.py (hunk 8 above)
Comment (counkomol) [reply fnet/cli/train_model.py]: I don't believe it's necessary to use `pathlib`. It's just one option.

File: examples/predict.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
+import t4
+import os
+import tqdm
+import json
+import pandas as pd
+
+from fnet.cli.init import save_default_train_options
```
Comment (counkomol) [inline examples/predict.py]: First, why is this import needed? Also, my understanding is that this is supposed to be an example of the `fnet` Python API, yet this file does not use the Python API. Wouldn't a bash script be clearer since you use the command-line interface here? Also, keep in mind that the user can run `fnet init` on the command line to see example scripts of how the `fnet` module can be used.

File: examples/download_and_train.py (hunk 7 above)
Comment (gregjohnso) [reply examples/download_and_train.py]: Yes, because it shows what the command is that actually got run!

File: fnet/cli/predict.py (hunk 13 above)
Comment (gregjohnso) [reply fnet/cli/predict.py]: readability

File: examples/download_and_train.py [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+
+aics_pipeline = t4.Package.browse(
+    "aics/pipeline",
+    registry="s3://quilt-aics"
+)
+
+len(aics_pipeline['fov'])
+
+image_ids = [k for k in aics_pipeline['fov']][0:n_images_to_download]
```
[the remark this answers is not in the document]
Comment (gregjohnso) [reply examples/download_and_train.py]: Did not know that. Thanks

File: fnet/cli/predict.py (hunk 13 above)
Comment (counkomol) [reply fnet/cli/predict.py]: I think this falls into the category of opinion, and I object to changes to files that have no affect on functionality. As a reminder, users can always run `fnet init` and customize the training/predict scripts as they desire.

File: examples/download_and_train.py [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+###################################################
+# Download the 3D multi-channel tiffs via Quilt/T4
+###################################################
+
+gpu_id = 0
+
+n_images_to_download = 40
+train_fraction = 0.75
+
+image_save_dir = '{}/images/'.format(os.getcwd())
```
[the remark this answers is not in the document]
Comment (gregjohnso) [reply examples/download_and_train.py]: thanks

File: examples/download_and_train.py [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+prefs['dataset_train_kwargs'] = {'path_csv': data_save_path_train}
+prefs['dataset_val'] = 'fnet.data.MultiChTiffDataset'
+prefs['dataset_val_kwargs'] = {'path_csv': data_save_path_test}
+
+# This Fnet call will be updated as a python API becomes available
+
+with open(prefs_save_path, 'w') as fp:
+    json.dump(prefs, fp)
+
+command_str = 'python ../fnet/cli/train_model.py {} --gpu_ids {}'.format(prefs_save_path, gpu_id)
```
Comment (counkomol) [inline examples/download_and_train.py]: Why not run `fnet train`? Also, the way it is now requires that the user be in a specific directory relative to the script.

File: examples/predict.py [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+gpu_id = 0
+
+image_save_dir = '{}/images/'.format(os.getcwd())
+model_save_dir = '{}/model/'.format(os.getcwd())
+prefs_save_path = '{}/prefs.json'.format(os.getcwd())
+
+data_save_path_test = '{}/image_list_test.csv'.format(image_save_dir)
+
+command_str = 'python  ../fnet/cli/predict.py ./ ' \
```
Comment (counkomol) [inline examples/predict.py]: Similar for the train script, the user should probably call `fnet predict` instead. Also, I still would like to see some justification for creating a Python script that creates a bash script to be run in Python. Why not have this as a bash script?

File: fnet/cli/predict.py [hunk 19]
```diff
@@ -93,6 +93,7 @@ def item_from_dataset(
...
```
Comment (counkomol) [inline fnet/cli/predict.py]: In the future, I'd prefer if we do not arbitrarily add whitespaces to the files with no functional additions.

File: fnet/cli/train_model.py [hunk 20]
```diff
@@ -118,6 +118,9 @@ def main(args: Optional[argparse.Namespace] = None) -> None:
         train_options = json.load(fi)
     args.__dict__.update(train_options)
 
+    if not os.path.exists(args.path_save_dir):
```
Comment (counkomol) [inline fnet/cli/train_model.py]: Why is this additional necessary?

File: fnet/data/__init__.py [hunk 21]
```diff
@@ -1,6 +1,7 @@
-from fnet.data.bufferedpatchdataset import BufferedPatchDataset
-from fnet.data.tiffdataset import TiffDataset
-from fnet.data.fnetdataset import FnetDataset
+from .bufferedpatchdataset import BufferedPatchDataset
```
Comment (counkomol) [inline fnet/data/__init__.py]: Why were all imports changed to be relative? According to [this](https://www.python.org/dev/peps/pep-0008/#imports), absolute imports are recommended. Also, if we wanted to change this particular `__init__.py` file, then we should change *all* of them.

File: fnet/data/multichtiffdataset.py [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+class MultiChTiffDataset(FnetDataset):
+    """Dataset for multi-channel tiff files.
+
+    Currently assumes that images are loaded in ZCXY format
+
+    """
+
+    def __init__(self, dataframe: pd.DataFrame = None, path_csv: str = None,
+                 transform_signal=[transforms.normalize],
+                 transform_target=None):
```
Comment (counkomol) [inline fnet/data/multichtiffdataset.py]: Shouldn't the target also be normalized? As defaults, I think either both signal and target should be normalized or nothing should be done at all.

File: fnet/data/multichtiffdataset.py [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+
+        super().__init__(dataframe, path_csv, transform_signal, transform_target)
+
+        self.df['channel_signal'] = [int(ch)
```
Comment (counkomol) [inline fnet/data/multichtiffdataset.py]: What happens if the element is `nan` or` None`?

File: setup.py [hunk 24]
```diff
@@ -23,13 +23,17 @@
         'tifffile',
         'torch==1.0',
         'tqdm',
+        'scikit-image==0.15.0',
```
Comment (counkomol) [inline setup.py]: Could we unpin this?

File: setup.py [hunk 25]
```diff
@@ -23,13 +23,17 @@
...
     ],
     extras_require={
         'dev': [
             'flake8',
             'pylint',
             'pytest',
             'pytest-cov',
+        ],
+        'examples': [
+            't4==0.0.10',
```
Comment (counkomol) [inline setup.py]: I don't know much about `t4`, but please double check if this needs to be pinned.

File: .gitignore [hunk 26]
```diff
@@ -6,6 +6,9 @@ data/*/
 !data/csvs/
 data/csvs/*/
 saved_models/
+*.ipynb_checkpoints/
+examples/*
```
Comment (counkomol) [inline .gitignore]: Why should we ignore files in the `examples` directory? Isn't a main point of this PR to add files to this directory?

File: setup.py (hunk 25 above)
Comment (gregjohnso) [reply setup.py]: It's api is in development, so I think pinning it is a safe thing to do.

File: .gitignore (hunk 26 above)
Comment (gregjohnso) [reply .gitignore]: You're right, I want to ignore subdirectories of that folder.

File: examples/predict.py [hunk 27]
```diff
@@ -0,0 +1,21 @@
...
+
+gpu_id = 0
+
+image_save_dir = '{}/images/'.format(os.getcwd())
+model_save_dir = '{}/model/'.format(os.getcwd())
+prefs_save_path = '{}/prefs.json'.format(os.getcwd())
+
+data_save_path_test = '{}/image_list_test.csv'.format(image_save_dir)
+
+command_str = 'fnet predict ./ ' \
```
Comment (counkomol) [inline examples/predict.py]: Again, this implies your model is saved in the cwd. `model_save_dir` above is unused. This also means your data directory, `images`, is inside your model directory.

File: examples/download_and_train.py [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+df_test.to_csv(data_save_path_test, index=False)
+df_train.to_csv(data_save_path_train, index=False)
+
+################################################
+# Run the label-free stuff (dont change this)
+################################################
+
+save_default_train_options(prefs_save_path)
+
+with open(prefs_save_path, 'r') as fp:
```
Comment (heeler) [inline examples/download_and_train.py]: why read the preferences from disk right after they have been written?

File: examples/download_and_train.py [hunk 29]
```diff
@@ -0,0 +1,97 @@
...
+prefs['interval_checkpoint'] = 10000
+
+prefs['dataset_train'] = 'fnet.data.MultiChTiffDataset'
+prefs['dataset_train_kwargs'] = {'path_csv': data_save_path_train}
+prefs['dataset_val'] = 'fnet.data.MultiChTiffDataset'
+prefs['dataset_val_kwargs'] = {'path_csv': data_save_path_test}
+
+# This Fnet call will be updated as a python API becomes available
+
+with open(prefs_save_path, 'w') as fp:
```
Comment (heeler) [inline examples/download_and_train.py]: 1) write the file
2) read the file
3) change/augment the values
4) write the file again
why so many file accesses?

File: examples/predict.py [hunk 30]
```diff
@@ -0,0 +1,21 @@
+import os
+
+###################################################
+# Assume the user already ran download_and_train.py
+###################################################
+
+gpu_id = 0
+
+image_save_dir = '{}/images/'.format(os.getcwd())
```
Comment (heeler) [inline examples/predict.py]: # make this cleaner
cwd = os.getcwd()
image_save_dir = '{cwd}/images/'
...

File: fnet/cli/train_model.py (hunk 20 above)
Comment (heeler) [inline fnet/cli/train_model.py]: I don't think you need the if statement here. 
os.makedirs makes the directories if they don't exist. 
if they exist I don't think it has any effect.

File: fnet/data/multichtiffdataset.py [hunk 31]
```diff
@@ -0,0 +1,61 @@
...
+                                     for ch in self.df['channel_signal']]
+        self.df['channel_target'] = [int(ch)
+                                     for ch in self.df['channel_target']]
+
+        assert all(i in self.df.columns for i in [
+                   'path_tiff', 'channel_signal', 'channel_target'])
+
+    def __getitem__(self, index):
```
Comment (heeler) [inline fnet/data/multichtiffdataset.py]: what's the logic behind has_target?

File: fnet/data/multichtiffdataset.py [hunk 32]
```diff
@@ -0,0 +1,61 @@
...
+        im_out.append(im_tmp[:, element['channel_signal']])
+
+        if has_target:
+            im_out.append(im_tmp[:, element['channel_target']])
+
+        if self.transform_signal is not None:
+            for t in self.transform_signal:
+                im_out[0] = t(im_out[0])
+
+        if has_target and self.transform_target is not None:
```
Comment (heeler) [inline fnet/data/multichtiffdataset.py]: this bothers me: has_target should be a boolean 
None is a different creature I'm not sure how this would equate. 
[this could be lack of knowledge on my part]

File: fnet/data/multichtiffdataset.py [hunk 33]
```diff
@@ -0,0 +1,61 @@
...
+
+        # unsqueeze to make the first dimension be the channel dimension
+        im_out = [torch.unsqueeze(im, 0) for im in im_out]
+
+        return tuple(im_out)
+
+    def __len__(self):
+        return len(self.df)
+
+    def get_information(self, index: int) -> dict:
```
Comment (heeler) [inline fnet/data/multichtiffdataset.py]: I really like the argument typing! Do more :)

File: examples/predict.py (hunk 30 above)
Comment (heeler) [reply examples/predict.py]: (sorry I meant this as a comment it turned it into a H1 header, my bad)

File: examples/download_and_train.py (hunk 29 above)
Comment (counkomol) [reply examples/download_and_train.py]: This a Python script that runs fnet's command-line tools. The json write/read/write is a symptom of that. I stated before that a bash script or a Python script that avoids the command-line interface would be clearer, but the claim is that this is a placeholder that will be replaced.

File: fnet/data/multichtiffdataset.py (hunk 23 above)
Comment (counkomol) [reply fnet/data/multichtiffdataset.py]: Noted in #112.

File: fnet/data/multichtiffdataset.py (hunk 31 above)
Comment (counkomol) [reply fnet/data/multichtiffdataset.py]: The target can be NaN/None, which implies a dataset item that is not meant for training or evaluation.

File: fnet/data/multichtiffdataset.py (hunk 32 above)
Comment (counkomol) [reply fnet/data/multichtiffdataset.py]: `self.target_target is not None` should evaluate to a boolean.

File: README.md [hunk 34]
```diff
@@ -15,20 +15,26 @@ An nVIDIA graphics card with >10GB of ram (we have used an nVIDIA Titan X Pascal
 
 ## Installation
 ### Environment setup
-- Install [Miniconda](https://conda.io/miniconda.html) if necessary.
-- Create a Conda environment for the platform:
+- Note: We no longer provide Conda files, but do recommend installing in a clean python 3.6 environment.
```
[the remark this answers is not in the document]
Comment (counkomol) [reply README.md]: #114

File: fnet/cli/train_model.py (hunk 20 above)
Comment (counkomol) [reply fnet/cli/train_model.py]: The function apparently now has an `exist_ok` parameter (https://docs.python.org/3/library/os.html#os.makedirs). This would make it unnecessary to check for directory existence.

File: examples/download_and_train.py (hunk 28 above)
Comment (counkomol) [reply examples/download_and_train.py]: See below.

File: .gitignore [hunk 35]
```diff
@@ -6,6 +6,9 @@ data/*/
 !data/csvs/
```
[the remark this answers is not in the document]
Comment (counkomol) [reply .gitignore]: I think the idea was to ignore directories within `data/csvs/` but not files. Anyway, this is a remnant of the repo formerly being just a code dump. We can work on cleaning this up going forward.

File: examples/download_and_train.py [hunk 36]
```diff
@@ -0,0 +1,97 @@
...
+image_save_paths = ['{}/{}'.format(image_save_dir, image_id) for image_id in image_ids]
+
+for image_id, image_save_path in zip(image_ids, image_save_paths):
+    if os.path.exists(image_save_path):
+        continue
+
+    # We only do this because T4 hates our filesystem. It probably wont affect you.
+    try:
+        aics_pipeline['fov'][image_id].fetch(image_save_path)
+    except OSError:
```
[the remark this answers is not in the document]
Comment (counkomol) [reply examples/download_and_train.py]: The covered call is from a third-party package. It's hard to know what exactly can go wrong.

[the remark this answers is not in the document]
Comment (JacksonMaxfield) [reply examples/download_and_train.py]: I can kind of chime in here. It's not really our filesystem so much as this script was ran on the DGX. The DGX apparently has an older linux kernel and because of that a flag needs to be set for users to be allowed to store metadata on files. Something like `xattrs=True`, so this check is actually a decent method for handling when certain files have metadata attached and some not depending on filesystem / OS.
